tests/functional/test_discovery.py

```diff
@@ -0,0 +1,371 @@
+        discovered_endpoint = "https://discovered.domain"
+        config = Config(endpoint_discovery_enabled=True)
+        client, http_stubber = self.create_client(
+            service_name="test-discovery-endpoint", config=config
+        )
+        with http_stubber as stubber:
+            self.add_describe_endpoints_response(stubber, discovered_endpoint)
+            client.test_discovery_required(Foo="bar")
+            self.assert_endpoint_discovery_used(stubber, discovered_endpoint)
+
+    def test_endpoint_discovery_required_with_discovery_disabled(self):
+        discovered_endpoint = "https://discovered.domain"
+        config = Config(endpoint_discovery_enabled=False)
+        client, http_stubber = self.create_client(
+            service_name="test-discovery-endpoint", config=config
+        )
+        self.add_describe_endpoints_response(http_stubber, discovered_endpoint)
+        with self.assertRaises(EndpointDiscoveryRequired):
+            client.test_discovery_required(Foo="bar")
+
+    def test_endpoint_discovery_required_with_custom_endpoint(self):
+        endpoint = "https://custom.domain/"
+        client, http_stubber = self.create_client(
+            service_name="test-discovery-endpoint", endpoint_url=endpoint
+        )
+        with http_stubber as stubber:
+            stubber.add_response(status=200, body=b'{}')
+            client.test_discovery_required(Foo="bar")
+            self.assert_discovery_skipped(
+                stubber,
+                b"test-discovery-endpoint.TestDiscoveryRequired"
+            )
+            self.assert_endpoint_used(stubber.requests[0].url, endpoint)
+
+    def test_endpoint_discovery_disabled_with_custom_endpoint(self):
+        endpoint = "https://custom.domain/"
+        config = Config(endpoint_discovery_enabled=False)
+        client, http_stubber = self.create_client(
+            service_name="test-discovery-endpoint",
+            config=config,
+            endpoint_url=endpoint
+        )
+        with http_stubber as stubber:
+            stubber.add_response(status=200, body=b'{}')
+            client.test_discovery_required(Foo="bar")
+            self.assert_discovery_skipped(
+                stubber,
+                b"test-discovery-endpoint.TestDiscoveryRequired"
+            )
+            self.assert_endpoint_used(stubber.requests[0].url, endpoint)
+
+    def test_endpoint_discovery_enabled_with_custom_endpoint(self):
+        endpoint = "https://custom.domain/"
+        config = Config(endpoint_discovery_enabled=True)
+        client, http_stubber = self.create_client(
+            service_name="test-discovery-endpoint",
+            config=config,
+            endpoint_url=endpoint
+        )
+        with http_stubber as stubber:
+            stubber.add_response(status=200, body=b'{}')
+            client.test_discovery_required(Foo="bar")
+            self.assert_discovery_skipped(
```

Comment:
This is a requirement for the feature, if we specify an endpoint it needs to be used as the final endpoint.

Reply:
That's fair. Thinking it over more, it would definitely be even more strange if the SDK actually tried endpoint discovery and ended up using a different endpoint than what was specified. Also there is no purpose to doing endpoint discovery if you already know what endpoint you want to use and explicitly specify it.